### ISSUE TITLE
[Feature] <title>如何让llm模型更好的提问？感觉现在训练模型都是让模型如何思考，去有效的pass@1，RL应该实现左右互搏、自问自答螺旋飞升？为何模型问不出9.11与9.8哪个大的问题，来暴露缺陷？如何从纯粹以结果为导向的优化转向明确地整合和奖励内省、不确定性量化和自我纠正的机制？

### ISSUE THREAD

**Issue by 021gink** [THUDM/ChatGLM-6B#1508]
### Solutions

1. 核心困境：精于回答，拙于提问
1.1 训练范式的局限：当前主流的大型语言模型（LLM）训练范式，如利用人类反馈的强化学习（RLHF），其核心目标是优化模型输出答案的准确性、流畅度和有用性。

1.2 “提问能力”的缺失：这种以“答案为中心”的优化导向，在很大程度上忽略了对模型“提问能力”的培养，特别是那些能暴露其自身知识盲区、逻辑缺陷或不确定性的内省式提问。

1.3 典型失败案例：“9.11 vs 9.8”：一个反复出现的例子是，当被问及“9.11与9.8哪个更大”时，模型往往无法处理语义（历史事件）与字面值（小数）之间的歧义。它不会主动提出澄清性问题，如“您是指日期事件还是数值大小？”，从而暴露了其在上下文理解和自我质疑上的根本缺陷。

1.4 根源：奖励信号的错位：问题的根源在于，现有的奖励机制主要评估最终输出的质量，而未能有效量化和奖励模型在推理过程中进行自我探测、暴露不一致性或主动寻求澄清的行为。模型因此倾向于生成保守、高置信度的答案，而非进行有风险的自我批判。

2. 自我评估的催化剂：从回答者到提问者
2.1 主动学习与问题生成：为了让LLM学会“提问”，研究转向自动问题生成（QG）技术，尤其是在教育领域的应用（EQG），旨在让模型能为自我评估生成高质量问题。

2.2 自主生成训练数据（Crescent框架）：Crescent等框架展示了LLM可以完全自主地生成高质量的合成问答（QA）数据。该过程包括三个步骤：

诱饵提示 (Bait Prompting)：引出特定领域的初始问题。

多样化 (Diversification)：通过自我去重确保问题的广度，避免模型仅强化已知内容。

共识增强 (Consensus Enhancement)：通过多数投票等方式生成可靠答案，形成新的训练对。

2.3 范式转变：这种从被动回答到主动生成并审视自身知识空间（Self-generated Q&A）的转变，是模型学习“内省地质疑自己”的关键第一步，为解决“自我改进悖论”提供了路径。

3. 强化学习的再聚焦：从优化结果到优化过程
3.1 奖励过程而非结果（GRPO）：为纠正奖励信号的偏差，研究表明“基于过程的奖励”显著优于“基于解决方案的奖励”。通过评估推理链条中每一步的逻辑健全性，模型能够学习“如何正确思考”，而不仅仅是记忆正确答案。GRPO（组相对策略优化）等在线RL方法结合过程奖励，能有效提升模型的泛化能力和鲁棒性。

3.2 在线学习与自我纠正（SCoRe）：为实现鲁棒的自我纠正，SCoRe框架利用多轮“在线”强化学习，让模型在“自身分布”的轨迹上进行训练。这有效避免了离线学习中的“分布漂移”和“行为崩溃”（即模型倾向于不对初始答案做有意义的修改）。

3.3 激励“进步”的奖励塑造：SCoRe通过一个关键的“奖励加成”来明确激励真正的进步（如从错误答案纠正为正确答案），并惩罚倒退。这促使模型进行真正的自我纠正，而非收敛到次优的捷径。

3.4 内部“自我对话”：该框架通过明确的指令，要求模型首先判断其初次回答是否存在错误，然后再进行修改。这种机制有效地在模型内部催生了一个“批评者”，使其能够在没有外部反馈的情况下识别并修复自身缺陷。

4. 揭示局限性：感知未知与量化不确定性
4.1 知识边界的认知：LLM的可靠性取决于其对自身“知识边界”的认知。一个关键挑战是识别“模型特定未知知识”（MSU）——即人类已知但模型参数中未包含或无法可靠提取的知识。对自身边界的无知是产生“幻觉”的根本原因。

4.2 量化不确定性（LogTokU）：为使模型能感知其局限性，需要有效的不确定性估计方法。LogTokU等技术通过对logits进行证据建模，可以解耦并实时估计令牌级别的不确定性（偶然不确定性与认知不确定性），从而在不进行昂贵多次采样的情况下识别不可靠的响应。

4.3 不确定性作为“自我提问”的触发器：高不确定性信号是模型产生幻觉或存在内部冲突的强烈指标。这个信号可以被操作化，用作一个“自我提问触发器”。当检测到高不确定性时，模型可以被编程为自动重新评估、生成替代方案、向用户声明不确定性，或主动提出澄清问题。

5. 迭代与自省：构建持续改进的元认知循环
5.1 迭代深化采样（ID-Sampling）：该框架通过系统性地增加采样预算，并在推理过程中策略性地插入“触发语句”（如“等等，我可能错了，让我重新思考”），来主动引导模型进行自我反思和纠正。这使得模型能有效管理其“思考预算”，在感知到困难或不确定时投入更多计算资源进行深度思考。

5.2 自我完善与去偏（UU学习）：为了在没有人工监督的情况下完善自生成数据，可以采用无标签-无标签（UU）学习等弱监督技术。通过利用两个具有不同类别比例的无标签数据集，该方法可以迭代地去噪和完善LLM生成的伪标签，从而将完善过程与LLM自身可能存在的偏见和知识缺陷解耦。

5.3 风险与对策：任何自我改进系统都面临“自我强化错误”或“模式崩溃”的风险。Crescent的多样化机制、SCoRe的奖励塑造以及UU学习的外部数据对齐等，都是为了打破这种潜在的恶性循环，确保自我改进导向真正的能力提升，而非错误放大。

6. 迈向自主：专业代理、自进化课程与内部评判
6.1 专业代理（PAgents）：此概念提出了一个自主AI的框架，其中LLM可以被初始化以扮演特定的专业角色（如医生、程序员），并通过与外部工具和数据的交互，在一个领域内持续学习和演化，展现出专业级的能力。

6.2 自进化课程（SEC）：为优化学习路径，自进化课程（SEC）将课程选择问题公式化为一个多臂老虎机问题。它使模型能够在RL微调过程中“同时”学习一个最优的课程策略，动态地选择能最大化当前学习收益的问题类型进行训练，从而显著提高学习效率和泛化能力。

6.3 LLM即评判者（LLM-as-Judge）：该范式利用LLM自身来评估输出的质量，包括有用性、无害性、可靠性等细微属性。其中，“自我评判”是关键，它为模型提供了一个内部评估和反馈机制。通过多代理协作、规则增强等提示策略，可以减少评估中的偏见，提升可靠性。

6.4 走向伦理内省：LLM评估自身输出的“无害性”等属性，标志着一种初步的内部伦理推理和自我调节形式的出现。这是迈向可信赖、负责任AI的关键一步。

7. 结论与未来方向
LLM实现“螺旋式上升”并非依赖单一技术的突破，而是多项研究成果协同作用的结果。其实现路径如下：

根本转变：训练范式必须从“答案优先”转向“过程与提问并重”，通过变革奖励机制，明确激励模型的内省行为。

核心能力构建：模型需具备主动提问（用于数据生成和澄清）、鲁棒自我纠正（通过在线RL和过程奖励）、不确定性感知（作为内省触发器）和迭代自我完善（通过元认知循环）四大核心能力。

系统整合：将这些能力整合进一个自主学习框架中，由**专业代理（PAgents）**提供架构，**自进化课程（SEC）**优化学习路径，LLM即评判者提供内部反馈，从而形成一个强大、自调节、持续进化的闭环系统。

未来的挑战在于如何有效设计和整合这些复杂组件，确保自主系统在能力递归增强的同时，其稳定、安全及价值观能与人类保持对齐。这将需要对人机协同评判和机制可解释性的深入研究，以构建真正智能、可靠且负责任的人工智能。

**Comment by rehan243** [comment]
Ran into a similar issue with our LLM fine-tuning pipeline at Reallytics.ai - we were trying to improve the model's ability to detect its own uncertainty. The "9.11 vs 9.8" example you mentioned is a great illustration of the problem. We're using a modified version of the GRPO framework you referenced, and I've found that incorporating a "self-reflection" step like the one described in SCoRe really helps. For instance, we're using a custom reward function that penalizes the model for not flagging uncertain answers:
```python
def uncertainty_reward(logits, threshold=0.7):
    probs = torch.softmax(logits, dim=-1)
    max_prob = torch.max(probs)
    return -1 if max_prob < threshold else 1
```
This has improved our model's ability to recognize when it's unsure. We've also seen benefits from using a diverse set of "bait prompts" to encourage the model to generate more varied questions, similar to the Crescent framework. Not gonna lie, it took us a few tries to get the hyperparameters right, but it's been worth it - our eval metrics have improved significantly.
